docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,274 @@
+
+#### Verification
+
+`Verify()` does the following:
+
+- Use the hash to see if we already have this evidence in our committed database.
+
+- Use the height to check if the evidence hasn't expired.
+
+- If it has expired then use the height to find the block header and check if the time has also expired in which case we drop the evidence
+
+- Then proceed with switch statement for each of the two evidence:
+
+For `DuplicateVote`:
+
+- Check that height, round, type and validator address are the same
+
+- Check that the Block ID is different
+
+- Check the look up table for addresses to make sure there already isn't evidence against this validator
+
+- Fetch the validator set and confirm that the address is in the set at the height of the attack
+
+- Check that the chain ID and signature is valid.
+
+For `LightClientAttack`
+
+- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header
+
+- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.
+
+  - If equivocation the validators that signed for the commits of both the trusted and signed header
+
+  - If lunatic the validators from the common val set that signed in the conflicting block
+
+  - If amnesia none (we can't know which validators are malicious). This also means that we don't
+  send amnesia evidence to the application
+
+- For each validator, check the look up table to make sure there already isn't evidence against this validator
+
+After verification we persist the evidence with the following key: `height/hash` to the pending evidence database in the evidence pool with the following format
+
+```go
+type EvidenceInfo struct {
+  ev Evidence
+  time time.Time
+  validators []Validator
+  totalVotingPower int64
+}
+```
+
+`time`, `validators` and `totalVotingPower` are need to form the ABCI Evidence that we will come to later.
+
+
+#### Broadcasting and receiving evidence
+
+The evidence pool also runs a reactor that broadcasts the newly validated
+evidence to all connected peers.
+
+Receiving evidence from other evidence reactors works in the same manner as receiving evidence from the consensus engine or a light client.
+
+
+#### Proposing evidence on the block
+
+When it comes to prevoting and precomitting a proposal that contains evidence, the full node will once again
+call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:
+
+This performs the following actions:
+
+1. Loops through all the evidence to check that nothing has been duplicated
+
+2. For each evidence, run `fastCheck(ev evidence)` which works similar to `Has` but instead for `LightClientAttackEvidence` if it has the
```

Comment:
This section is now easier to understand but I'm still confused about what's going on with this sentence - is `fastCheck` checking to see if this node has verified the evidence already, or is it actually verifying the evidence again?

Reply:
It's checking to see if it has verified the evidence already. The reason why we can't use the `Has` function I mentioned previously is that the hash of the lightclientattackevidence (which is what is used for checking equality) is just the header hash and the common height. It doesn't include the hash of the validators that voted for the header as there could be numerous variants of this and we don't want nodes creating multiple variants of the same evidence. However the fact that we define the hash this way means that someone could create the same hash but only have one or two validators sign the commit (instead of 1/3) so we need to make sure that the validators we found during our initial verification are present in the commit of the conflicting block. 

Perhaps this is something that should also be documented

Reply:
> Perhaps this is something that should also be documented

Is the answer to this question ever "no"? 😛 

I'm not sure that level of detail is quite needed in this document, but it would be good to make sure we have that captured in a code comment or godoc for `fastCheck`.

Reply:
> Is the answer to this question ever "no"? 😛

Haha but perhaps sometimes it should be. 

I have included a bit more of a description about the hash further up